### PR TITLE
fix: server action cases

### PR DESCRIPTION
**Pull request by himself65**
Fix some RSC case, will fix https://github.com/dai-shi/waku/pull/745. In the future, I will migrate the logic to https://github.com/himself65/react-server and keep same output

Some issues still happening:

1. `export default async function () {"use server"}` cannot be RSCA
2. `export function a() { "use server" }` should throw error, because it's not async

### CODE REVIEW

File: packages/waku/tests/vite-plugin-rsc-transform-internals.test.ts [hunk 1]
```diff
@@ -68,6 +68,81 @@ export const log = (mesg) => {
...
+    wrappedActions[name] = wrapAction(actions[name]);
+  }
+  return function AI(props) {
+    return jsx(InternalProvider, {
+      actions: wrappedActions,
+      children: props.children,
+    });
+  };
+}`;
+    // fixme: this is wrong, `innerAction.bind` should be `__waku_action1.bind`
```
Comment (himself65) [inline packages/waku/tests/vite-plugin-rsc-transform-internals.test.ts:116]: /cc @dai-shi

File: packages/waku/tests/vite-plugin-rsc-transform-internals.test.ts [hunk 2]
```diff
@@ -68,10 +66,117 @@ export const log = (mesg) => {
...
+  });
+
+  test('create server with bind', async () => {
+    const code = `
+'use server';
+import { InternalProvider } from './shared.js';
+import { jsx } from 'react/jsx-runtime';
+
+async function innerAction({ action }, ...args) {
+  'use server';
```
Comment (dai-shi) [inline packages/waku/tests/vite-plugin-rsc-transform-internals.test.ts:120]: Okay, you want to support this double `use server` case. I thought about it, but I was not sure if it's valid.

File: packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts [hunk 3]
```diff
@@ -417,44 +417,96 @@ export ${name === 'default' ? name : `const ${name} =`} registerClientReference(
...
+      }
+      let useServerPosStart: number;
+      let useServerPosEnd: number;
+      if (hasUseServer === true) {
+        useServerPosStart = 0;
+        useServerPosEnd = 0;
+      } else {
+        useServerPosStart = hasUseServer.span.start - mod.span.start;
+        useServerPosEnd = hasUseServer.span.end - mod.span.start;
+      }
```
Comment (dai-shi) [inline packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts:481]: The change above seems a bit ugly.
Let me see if I can do anything with it.

**Review by dai-shi — COMMENTED** [review]
Thanks for the catch.

File: packages/waku/tests/vite-plugin-rsc-transform-internals.test.ts (hunk 2 above)
Comment (himself65) [reply packages/waku/tests/vite-plugin-rsc-transform-internals.test.ts:120]: It's valid in Next.js I think🤔

Comment (himself65) [reply packages/waku/tests/vite-plugin-rsc-transform-internals.test.ts]: I wrote down all cases in sheet
image

Comment (dai-shi) [reply packages/waku/tests/vite-plugin-rsc-transform-internals.test.ts]: yeah, it's reasonable with the object param case.

File: packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts (hunk 3 above)
Comment (dai-shi) [reply packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts]: done.

**Comment by dai-shi** [comment]
> 1. `export default async function () {"use server"}` cannot be RSCA

Yeah, I notice that too. Even with top-level "use server", it fails. Gotta fix it, but maybe later.

> 2. `export function a() { "use server" }` should throw error, because it's not async

Originally, I thought sync functions are supported, and I have such an example in 23_actions.
Does nextjs throw an error? I'll leave it to your `react-server`.

**Comment by himself65** [comment]
> Originally, I thought sync functions are supported, and I have such an example in 23_actions.
Does nextjs throw an error? I'll leave it to your react-server.

It must be an async (and next.js forces this) because the server action is function between the client and server, the is no way to be sync

**Comment by dai-shi** [comment]
> > Originally, I thought sync functions are supported, and I have such an example in 23_actions.
> > Does nextjs throw an error? I'll leave it to your react-server.
> 
> It must be an async (and next.js forces this) because the server action is function between the client and server, the is no way to be sync

#778

**Pull request merged**